fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexAggregate.java

```diff
@@ -0,0 +1,335 @@
+        @Nonnull
+        private final IndexAggregateGroupKeys groupKeys;
+        @Nonnull
+        private final String indexName;
+
+        IndexNode(@Nonnull QueryComponent filter, @Nonnull IndexAggregateGroupKeys groupKeys, @Nonnull String indexName) {
+            this.filter = filter;
+            this.groupKeys = groupKeys;
+            this.indexName = indexName;
+        }
+    }
+
+    static class Builder {
+        @Nonnull
+        private final FDBRecordStore recordStore;
+        @Nonnull
+        private final List<String> recordTypeNames;
+        @Nonnull
+        private final List<QueryComponent> groupFilters;
+        @Nonnull
+        private final IndexAggregateFunction indexAggregateFunction;
+        @Nullable
+        private Map<KeyExpression, Index> bitmapIndexes;
+        @Nullable
+        private Map<QueryComponent, IndexNode> indexNodes;
+
+        Builder(@Nonnull final FDBRecordStore recordStore, @Nonnull List<String> recordTypeNames,
+                @Nonnull List<QueryComponent> groupFilters, @Nonnull IndexAggregateFunction indexAggregateFunction) {
+            this.recordStore = recordStore;
+            this.recordTypeNames = recordTypeNames;
+            this.groupFilters = groupFilters;
+            this.indexAggregateFunction = indexAggregateFunction;
+        }
+
+        @Nonnull
+        Optional<Node> tryBuild(@Nonnull QueryComponent indexFilter) {
+            if (indexFilter instanceof ComponentWithComparison) {
+                return indexScan(indexFilter);
+            }
+            if (indexFilter instanceof AndOrComponent) {
+                final AndOrComponent andOrComponent = (AndOrComponent) indexFilter;
+                List<Node> childNodes = new ArrayList<>(andOrComponent.getChildren().size());
+                for (QueryComponent child : andOrComponent.getChildren()) {
+                    Optional<Node> childNode = tryBuild(child);
+                    if (!childNode.isPresent()) {
+                        return Optional.empty();
+                    }
+                    childNodes.add(childNode.get());
+                }
+                final OperatorNode.Operator operator = indexFilter instanceof OrComponent ? OperatorNode.Operator.OR : OperatorNode.Operator.AND;
+                return Optional.of(new OperatorNode(operator, childNodes));
+            }
+            if (indexFilter instanceof NotComponent) {
+                return tryBuild(((NotComponent) indexFilter).getChild())
+                        .map(childNode -> new OperatorNode(OperatorNode.Operator.NOT, Collections.singletonList(childNode)));
+            }
+            return Optional.empty();
+        }
+
+        @Nonnull
+        Optional<Node> indexScan(@Nonnull QueryComponent indexFilter) {
+            if (bitmapIndexes == null) {
+                bitmapIndexes = findBitmapIndexes(indexAggregateFunction.getName());
+                indexNodes = new HashMap<>();
```

Comment:
I suppose the construction of `indexNodes` should be delayed until after `bitmapIndexes.isEmpty()` is checked... Not that I expect that change will make any, like, measurable difference in performance

Reply:
Sure. Not much of a difference, it's it's only a matter of reordering the statements.